support systemd

dffischer opened the pull request:
While [packaging this module for Arch Linux](https://aur.archlinux.org/packages/nodejs-vinefeed/), I made vinefeed play nicely with systemd. Incorporation of these changes here will make users of other distributions benefit, too.

This pull request provides unit files to start it as a system or user service and adds a few lines on how to use them to the `README`. Socket activation is also supported so that the program will only be started up when its port is accessed.

dffischer committed:
Support socket activation through systemd

When run as a standalone server, it can be helpful to have vinefeed
activated on demand.

dffischer committed:
Declare binary

This makes it globally accessible after installation through npm.

dffischer committed:
Provide systemd units

They can be used to start the service on port access.

tremby commented on README.md:
```diff
@@ -12,6 +12,15 @@ Setting up
 4. Go to `localhost:3000/users/<vine-user-id>/atom` or 
```
Can you add to item 3 here something like "or use systemd; see below"?

tremby commented on README.md:
```diff
@@ -12,6 +12,15 @@ Setting up
 4. Go to `localhost:3000/users/<vine-user-id>/atom` or 
    `localhost:3000/users/<vine-user-id>/rss`
 
+Usage with systemd
```
Can you add a note to item 3 above (under "setting up") to mention systemd? Maybe just ", or use systemd; see below".

dffischer replied:
Done.

tremby commented on README.md:
```diff
@@ -12,6 +12,15 @@ Setting up
+Usage with systemd
+------------------
+
+1. Copy `vinefeed.service` and `vinefeed.socket` to a systemd units directory, e.g. `/usr/lib/systemd/system` or `~/.config/systemd/user`
```
Maybe "Copy or symlink", if symlinks would work too? I realize it's a bad idea to symlink if it's going to be root running it though.

dffischer replied:
I assume a symlink would be fine, especially when only used in the user instance. However, the binary has to be available at `/usr/bin/vinefeed`, so I noted about installation, too.

tremby commented on vinefeed.socket:
```diff
@@ -0,0 +1,6 @@
+
+[Install]
+WantedBy=sockets.target
+
```
Kill the blank line

dffischer replied:
Done.

tremby commented:
I'm not likely to test this one, but I'm happy to merge it after you look at the above comments if you say it's working. (And as mentioned in the other ticket, please capitalize your commit messages.)

tremby commented on bin/www:
```diff
@@ -13,5 +13,5 @@ app.set 'port', process.env.PORT
 console.log "Environment: #{app.get('env')}"
 
-server = app.listen app.get('port'), ->
+server = app.listen require('systemd-socket')() || app.get('port'), ->
```
I'm not sure exactly what this does. Does this make the systemd-socket dependency optional? If so, perhaps the systemd-socket dependency in package.json should be moved to [optionalDependencies](https://docs.npmjs.com/files/package.json#optionaldependencies) and a note added to the readme about this.

Can the package still be run on a port after doing this? Does the user need further instruction now on how to configure for a port or a socket?

dffischer replied:
The dependency originally was not optional, but now made so with a try/catch.

The `systemd-socket` package automatically detects whether the application was run from systemd and acts accordingly. When it does not respond to socket activation, it uses the port exactly like it did without the change.

dffischer committed:
Make systemd-socket dependency optional

A missing systemd-socket package can simply be handled by always acing
as if the program was invoked directly, restoring the behaviour before
the dependency was even introduced.

tremby commented on bin/www:
```diff
@@ -1,5 +1,8 @@
 #!/usr/bin/env coffee
 
+try systemdsocket = require 'systemd-socket'
+catch e then systemdsocket = -> null
```
Really nitpicking here, but can you expand this to more lines:

```
try
    systemdsocket = require 'systemd-socket'
catch e
    systemdsocket = -> null
```

dffischer replied:
Feel free to nitpick as much as you like. Done.

tremby commented on README.md:
```diff
@@ -8,10 +8,22 @@ Setting up
+Usage with systemd
+------------------
+
+1. Globally install vinefeed with `npm install -g`
```
Should there be a note that the optional dependency needs to be installed too?

dffischer replied:
I do not think so. Omitting the optional dependencies would require an explicit `--no-optional` switch, so the command given here already installs them by default.

tremby replied:
Ah, I see.

dffischer commented:
I also rebased this branch onto the new master for easier merging of a linear history.

dffischer commented:
Thanks for the merge. This simplifies packaging a lot.
